File: pydss/meter/MeterElementImpl.py
from dss.common.impl.Utilities import (Utilities,)
from dss.common.impl.DSSGlobals import (DSSGlobals,)
from dss.meter.MeterElement import (MeterElement,)
from dss.common.impl.DSSClassDefs import (DSSClassDefs,)
from dss.common.impl.DSSCktElement import (DSSCktElement,)
# from org.apache.commons.math.complex.Complex import (Complex,)


class MeterElementImpl(DSSCktElement, MeterElement):
    elementName = None
    meteredElement = None
    meteredTerminal = None
    meteredElementChanged = None
    sensorCurrent = None
    sensorVoltage = None
    phsAllocationFactor = None
    calculatedCurrent = None
    calculatedVoltage = None
    avgAllocFactor = None

    def __init__(self, parClass):
        super(MeterElementImpl, self)(parClass)
        self.objType = DSSClassDefs.METER_ELEMENT
        self.elementName = ''
        self.meteredElement = None
        self.meteredTerminal = 1
        self.sensorCurrent = None
        self.sensorVoltage = None
        self.phsAllocationFactor = None
        self.calculatedCurrent = None
        self.calculatedVoltage = None

    def allocateSensorArrays(self):
        if self.meteredElement is not None:
            self.calculatedCurrent = Utilities.resizeArray(self.calculatedCurrent, self.meteredElement.getYorder())
        if self.meteredElement is not None:
            self.calculatedVoltage = Utilities.resizeArray(self.calculatedVoltage, self.meteredElement.getYorder())
        self.sensorCurrent = Utilities.resizeArray(self.sensorCurrent, self.nPhases)
        self.sensorVoltage = Utilities.resizeArray(self.sensorVoltage, self.nPhases)
        self.phsAllocationFactor = Utilities.resizeArray(self.phsAllocationFactor, self.nPhases)

    def calcAllocationFactors(self):
        self.meteredElement.getCurrents(self.calculatedCurrent)
        # the phase allocation factor is the amount that the load must change to match the measured peak
        iOffset = (self.meteredTerminal - 1) * self.meteredElement.getNConds()
        self.avgAllocFactor = 0.0
        _0 = True
        i = 0
        while True:
            if _0 is True:
                _0 = False
            else:
                i += 1
            if not (i < self.nPhases):
                break
            mag = self.calculatedCurrent[i + iOffset].abs()
            if mag > 0.0:
                self.phsAllocationFactor[i] = self.sensorCurrent[i] / mag
            else:
                self.phsAllocationFactor[i] = 1.0
                # no change
            self.avgAllocFactor = self.avgAllocFactor + self.phsAllocationFactor[i]
        self.avgAllocFactor = self.avgAllocFactor / self.nPhases
        # factor for 2- and 3-phase loads

    def takeSample(self):
        """Sample control quantities and set action times in control queue."""
        DSSGlobals.doSimpleMsg('Programming error: Reached base MeterElement class for takeSample.' + DSSGlobals.CRLF + 'Device: ' + self.getName(), 723)

    def getElementName(self):
        return self.elementName

    def setElementName(self, name):
        self.elementName = name

    def getMeteredElement(self):
        return self.meteredElement

    def setMeteredElement(self, element):
        self.meteredElement = element

    def getMeteredTerminal(self):
        return self.meteredTerminal

    def setMeteredTerminal(self, terminal):
        self.meteredTerminal = terminal

    def isMeteredElementChanged(self):
        return self.meteredElementChanged

    def setMeteredElementChanged(self, changed):
        self.meteredElementChanged = changed

    def getSensorCurrent(self):
        return self.sensorCurrent

    def setSensorCurrent(self, current):
        self.sensorCurrent = current

    def getSensorVoltage(self):
        return self.sensorVoltage

    def setSensorVoltage(self, voltage):
        self.sensorVoltage = voltage

    def getPhsAllocationFactor(self):
        return self.phsAllocationFactor

    def setPhsAllocationFactor(self, factor):
        self.phsAllocationFactor = factor

    def getCalculatedCurrent(self):
        return self.calculatedCurrent

    def setCalculatedCurrent(self, current):
        self.calculatedCurrent = current

    def getCalculatedVoltage(self):
        return self.calculatedVoltage

    def setCalculatedVoltage(self, voltage):
        self.calculatedVoltage = voltage

    def getAvgAllocFactor(self):
        return self.avgAllocFactor

    def setAvgAllocFactor(self, factor):
        self.avgAllocFactor = factor

File: core/jdss/src/main/java/com/epri/dss/meter/impl/MeterElementImpl.java
package com.epri.dss.meter.impl;

import org.apache.commons.math.complex.Complex;

import com.epri.dss.common.CktElement;
import com.epri.dss.common.DSSClass;
import com.epri.dss.common.impl.DSSCktElement;
import com.epri.dss.common.impl.DSSClassDefs;
import com.epri.dss.common.impl.DSSGlobals;
import com.epri.dss.common.impl.Utilities;
import com.epri.dss.meter.MeterElement;

public class MeterElementImpl extends DSSCktElement implements MeterElement {

	protected String elementName;
	protected CktElement meteredElement;
	protected int meteredTerminal;
	protected boolean meteredElementChanged;

	protected double[] sensorCurrent;
	protected double[] sensorVoltage;
	protected double[] phsAllocationFactor;
	protected Complex[] calculatedCurrent;
	protected Complex[] calculatedVoltage;
	protected double avgAllocFactor;

	public MeterElementImpl(DSSClass parClass) {
		super(parClass);
		objType = DSSClassDefs.METER_ELEMENT;

		elementName         = "";
		meteredElement      = null;
		meteredTerminal     = 0;
		sensorCurrent       = null;
		sensorVoltage       = null;
		phsAllocationFactor = null;
		calculatedCurrent   = null;
		calculatedVoltage   = null;
	}

	public void allocateSensorArrays() {
		if (meteredElement != null)
			calculatedCurrent = Utilities.resizeArray(calculatedCurrent, meteredElement.getYorder());
		if (meteredElement != null)
			calculatedVoltage = Utilities.resizeArray(calculatedVoltage, meteredElement.getYorder());
		sensorCurrent = Utilities.resizeArray(sensorCurrent, nPhases);
		sensorVoltage = Utilities.resizeArray(sensorVoltage, nPhases);
		phsAllocationFactor = Utilities.resizeArray(phsAllocationFactor, nPhases);
	}

	public void calcAllocationFactors() {
		int iOffset;
		int i;
		double mag;

		meteredElement.getCurrents(calculatedCurrent);

		// the phase allocation factor is the amount that the load must change to match the measured peak
		iOffset = meteredTerminal * meteredElement.getNConds();
		avgAllocFactor = 0.0;
		for (i = 0; i < nPhases; i++) {
			mag = calculatedCurrent[i + iOffset].abs();
			if (mag > 0.0) {
				phsAllocationFactor[i] = sensorCurrent[i] / mag;
			} else {
				phsAllocationFactor[i] = 1.0;  // no change
			}
			avgAllocFactor = avgAllocFactor + phsAllocationFactor[i];
		}
		avgAllocFactor = avgAllocFactor / nPhases;  // factor for 2- and 3-phase loads
	}

	/**
	 * Sample control quantities and set action times in control queue.
	 */
	public void takeSample() {
		DSSGlobals.doSimpleMsg("Programming error: Reached base MeterElement class for takeSample."+DSSGlobals.CRLF+"Device: "+getName(), 723);
	}

	public String getElementName() {
		return elementName;
	}

	public void setElementName(String name) {
		elementName = name;
	}

	public CktElement getMeteredElement() {
		return meteredElement;
	}

	public void setMeteredElement(CktElement element) {
		meteredElement = element;
	}

	public int getMeteredTerminal() {
		return meteredTerminal;
	}

	public void setMeteredTerminal(int terminal) {
		meteredTerminal = terminal;
	}

	public boolean isMeteredElementChanged() {
		return meteredElementChanged;
	}

	public void setMeteredElementChanged(boolean changed) {
		meteredElementChanged = changed;
	}

	public double[] getSensorCurrent() {
		return sensorCurrent;
	}

	public void setSensorCurrent(double[] current) {
		sensorCurrent = current;
	}

	public double[] getSensorVoltage() {
		return sensorVoltage;
	}

	public void setSensorVoltage(double[] voltage) {
		sensorVoltage = voltage;
	}

	public double[] getPhsAllocationFactor() {
		return phsAllocationFactor;
	}

	public void setPhsAllocationFactor(double[] factor) {
		phsAllocationFactor = factor;
	}

	public Complex[] getCalculatedCurrent() {
		return calculatedCurrent;
	}

	public void setCalculatedCurrent(Complex[] current) {
		calculatedCurrent = current;
	}

	public Complex[] getCalculatedVoltage() {
		return calculatedVoltage;
	}

	public void setCalculatedVoltage(Complex[] voltage) {
		calculatedVoltage = voltage;
	}

	public double getAvgAllocFactor() {
		return avgAllocFactor;
	}

	public void setAvgAllocFactor(double factor) {
		avgAllocFactor = factor;
	}

}
